### ISSUE TITLE
TB 115 not supported

### ISSUE THREAD

**Issue by Frank071** [TB-throwback/LookOut-fix-version#102]
This extension reports itself as not supporting TB 115 (SuperNova). Not sure if this is a matter of definition or if work needs to be done to make it support it. TB 115 was released to the public yesterday (July 11th).

**Comment by jobisoft** [comment]
I am looking at it.

**Comment by jobisoft** [comment]
@dugite-code : I created a wip-115 branch. It is not ready for a pull-request, but I wanted to share what I have done so far.

[lookout_fix_version-beta-tb.zip](https://github.com/TB-throwback/LookOut-fix-version/files/12043751/lookout_fix_version-beta-tb.zip)

This attempts to move away from the WindowListener API not fully done, options page still needs it) and use a dedicated Experiment for the attachment manipulation (done). I also moved the tnef stuff into the background and created a PseudoInputStream, which partially implements nsIInputStream in WebExtension land.

Could you test the current state? No options page so far, hardcoded in the background.

What do you think?

**Comment by IskrenTs** [comment]
I have tested the beta version on an old email i have and i can properly see the attachment files  just like on 102v of TB

**Comment by jobisoft** [comment]
Removed the WindowListener API, converted locales to WebExtension locales and converted options to html:

[lookout_fix_version-beta-tb.zip](https://github.com/TB-throwback/LookOut-fix-version/files/12044638/lookout_fix_version-beta-tb.zip)

**Comment by dugite-code** [comment]
That's amazingly quick work! A few notes I'm looking through:

## RangeError

I'm running through my old folder of TNEF emails and a few of them pop out the error: `RangeError: too many function arguments 3 lookout.mjs:49:19` I'm not 100% sure what's going on with these particular emails but the bytearray appears to be quite large.

## Old Preferences:
 ### strict_contenttype
I believe the strict_contenttype preference wouldn't be needed anymore as it appears to not necessarily derived from the emails stated content type? winmail.dat that wouldn't have been opened previously appear to open fine now.

I noticed the loop in [background.js#L31](https://github.com/TB-throwback/LookOut-fix-version/blob/wip-115/src/background.js#L31) appears to not get all the attachments the message has if the first attachment is an .eml file. Might be an edge case?

### direct_to_calendar
This preference was part of a dirty hack to have the ics file opened by the calendar. Currently when the attachments are updated it's not triggering the ice file to be detected, just as before.

### enable_debug
The enable debugging option in preferences obviously doesn't do anything at the moment and it's debatable if we want to add any logging back at the moment.

**Comment by jobisoft** [comment]
> I'm running through my old folder of TNEF emails and a few of them pop out the error: `RangeError: too many function arguments 3 lookout.mjs:49:19` I'm not 100% sure what's going on with these particular emails but the bytearray appears to be quite large.

Could you forward one of these messages to me?

> ## Old Preferences:
> ### strict_contenttype
> 
> I believe the strict_contenttype preference wouldn't be needed anymore as it appears to not necessarily derived from the emails stated content type? winmail.dat that wouldn't have been opened previously appear to open fine now.
>

The trick I am using is to check for filename "winmail.dat" and always try to parse it now. The option could still be used to enforce parsing of all attachments.

> 
> I noticed the loop in [background.js#L31](https://github.com/TB-throwback/LookOut-fix-version/blob/wip-115/src/background.js#L31) appears to not get all the attachments the message has if the first attachment is an .eml file. Might be an edge case?
>
There is a mode which extracts and display attachments of attached emails and I suppose those are not caught by the messages API. I plan to add a listAttachments method to the messageDisplay API, which always returns what the user sees (including manipulated attachments), that should fix that error. For now, we could manually dive into attached messages. 

> ### direct_to_calendar
> 
> This preference was part of a dirty hack to have the ics file opened by the calendar. Currently when the attachments are updated it's not triggering the ice file to be detected, just as before.

I skipped all that additional logic for now. Will check it out

> ### enable_debug
> 
> The enable debugging option in preferences obviously doesn't do anything at the moment and it's debatable if we want to add any logging back at the moment.
>

Hm, it should work. I will have a look.

> # Edit: tnef_parse and the msg_header
> 
> As per this todo [lookout.mjs#L65](https://github.com/TB-throwback/LookOut-fix-version/blob/wip-115/src/scripts/lookout.mjs#L65) the header is used to map names missing the email address to email addresses from the message header in event invites [tnef.mjs#L1064](https://github.com/TB-throwback/LookOut-fix-version/blob/wip-115/src/scripts/tnef.mjs#L1064).

Yeah, I also skipped that for the moment. The msgHdr passed into the TNEF stuff is only used to manipulate invites? I will check that. Could you forward me a message which would trigger this?

**Comment by jobisoft** [comment]
I think I fixed the debug flag issue and the issue where inline attachments of attached messages are not processed (this is indeed only happening if you enable "view inline attachments"). I added a listAttachment() and a getAttachmentFile() method, which returns the actually displayed attachments in the given tab.

I was not yet able to fix the parsing of the "Letto: c.a. dott.ssa" message, which logs:

```
TNEF: Entering tnef_parse() [tnef.mjs:132:13](moz-extension://df7a434f-b84e-4e3e-957e-400bdd9fda10/scripts/tnef.mjs)
TNEF: new TnefPackage() [tnef.mjs:132:13](moz-extension://df7a434f-b84e-4e3e-957e-400bdd9fda10/scripts/tnef.mjs)
TNEF Key: 0639 [tnef.mjs:132:13](moz-extension://df7a434f-b84e-4e3e-957e-400bdd9fda10/scripts/tnef.mjs)
checksum = 1  sum = 1 [tnef.mjs:132:13](moz-extension://df7a434f-b84e-4e3e-957e-400bdd9fda10/scripts/tnef.mjs)
checksum = 232  sum = 232 [tnef.mjs:132:13](moz-extension://df7a434f-b84e-4e3e-957e-400bdd9fda10/scripts/tnef.mjs)
TNEF: OEM Code Page = 1252 [tnef.mjs:132:13](moz-extension://df7a434f-b84e-4e3e-957e-400bdd9fda10/scripts/tnef.mjs)
checksum = 2819  sum = 2819 [tnef.mjs:132:13](moz-extension://df7a434f-b84e-4e3e-957e-400bdd9fda10/scripts/tnef.mjs)
checksum = 1955  sum = 1955 [tnef.mjs:132:13](moz-extension://df7a434f-b84e-4e3e-957e-400bdd9fda10/scripts/tnef.mjs)
checksum = 55370  sum = 55109 [tnef.mjs:132:13](moz-extension://df7a434f-b84e-4e3e-957e-400bdd9fda10/scripts/tnef.mjs)
LookOut: TNEF attribute has invalid checksum, message may be corrupt
```

There is also an issue with the method used to update the displayed attachments after we manipulated them, if we did not remove all original attachments. These original attachments then throw:

```
attachment.partID is null msgHdrView.js:2001
    getAttachmentsTotalSizeStr chrome://messenger/content/msgHdrView.js:2001
    updateAttachmentsDisplay chrome://messenger/content/msgHdrView.js:1961
    isEmpty resource:///modules/AttachmentInfo.sys.mjs:599
    AsyncFunctionNext self-hosted:852
```

Any idea?

**Comment by jobisoft** [comment]
While reactivating the remove_winmail_dat preference, I decided to remove all successfully parsed attachments, not only winmail.dat. Is that correct?

**Comment by jobisoft** [comment]
Can you confirm that the "Letto: c.a. dott.ssa" message works in 102? For me 102 and 115 behave exactly the same and both throw

```
checksum = 55370  sum = 55109
LookOut: TNEF attribute has invalid checksum, message may be corrupt
```

The only thing which works in both is to include the additional MAPI properties as attachment, but I do not see any real data.

**Comment by jobisoft** [comment]
Status:
 - icf stuff does not work
 - saving attachments does not fully work (but they are file attachemnts anyhow and the context menu takes you to the local temp file )

Anything else?

I would consider this state as releasable, do you agree?. Creating the PullRequest.

[lookout_fix_version-beta-tb.zip](https://github.com/TB-throwback/LookOut-fix-version/files/12049823/lookout_fix_version-beta-tb.zip)

**Comment by dugite-code** [comment]
> Can you confirm that the "Letto: c.a. dott.ssa" message works in 102? For me 102 and 115 behave exactly the same and both throw
> 
> ```
> checksum = 55370  sum = 55109
> LookOut: TNEF attribute has invalid checksum, message may be corrupt
> ```
> 
> The only thing which works in both is to include the additional MAPI properties as attachment, but I do not see any real data.

Ah sorry I somehow got a broken sample mixed into the working samples folder. You can ignore that one


> Status:
> 
>     * icf stuff does not work
> 
>     * saving attachments does not fully work (but they are file attachments anyhow and the context menu takes you to the local temp file)
> 
> 
> Anything else?
> 
> I would consider this state as releasable, do you agree? Creating the PullRequest.
> 
> [lookout_fix_version-beta-tb.zip](https://github.com/TB-throwback/LookOut-fix-version/files/12049823/lookout_fix_version-beta-tb.zip)

I think that'll be fine for a quick release. File drag and drop support is fixed so that's a nice bonus. 

> While reactivating the remove_winmail_dat preference, I decided to remove all successfully parsed attachments, not only winmail.dat. Is that correct?

My initial gut reaction was yes because we expect to mainly be dealing with Microsoft emails where there shouldn't be any additional attachments. That said I have one example where this isn't ideal, forwarded .eml files now have their winmail.dat files parsed in the parent email but the forwarded emails are removed eliminating any context. This is something we'll have to re-visit I think.

**Comment by dugite-code** [comment]
New version has been uploaded to ATN and I've created new issues for the regressions and the winmail.dat removal question.
